htmresearch/support/text_preprocess.py

```diff
@@ -147,50 +156,97 @@ def readExpansionFile(filename, suffixes=None):
     return expansionPairs
 
 
-  def tokenize(self,
-               text,
-               ignoreCommon=None,
-               removeStrings=None,
-               correctSpell=False,
-               expandAbbr=False,
-               expandContr=False):
+  def tokenizeAndFilter(self, text, **preprocessSpecs):
     """
-    Tokenize, returning only lower-case letters and "$".
-    @param text               (str)             Single string to tokenize.
-    @param ignoreCommon       (int)             This many most frequent words
-                                                will be filtered out from the
-                                                returned tokens.
-    @param removeStrings      (list)            List of strings to delete from
-                                                the text.
-    @param correctSpell       (bool)            Run tokens through spelling
-                                                correction.
-    @param expandAbbr         (bool)            Run text through abbreviation
-                                                expander
-    @param expandContr        (bool)            Run text through contraction
-                                                expander
+    Tokenize and optionally filter the text, returning both the tokens and a
+    mapping of the original text to the preprocessed text.
+
+    @param text            (str)    Text to be tokenized and preprocessed.
+
+    @return processedWords (list)   String tokens.
+
+    @return mapping        (dict)   Keys are the indices of the input words.
+        Values are two-tuples, where the first item is the original string word
+        (w/ punctuation) and the second item is a list of string tokens. This
+        format is necessary to account for when there are more tokens than words
+        and vice-versa, and can't use words as keys b/c repeats would get
+        overwritten.
     """
+    originalWords = text.split(" ")
+    processedWords = []
+    mapping  = OrderedDict()
```

Comment:
I thought it could be useful to have the order preserved, but it would be simpler to use a normal dict. If order is something we need later we can always change it then.